src/OpenIddict.Core/Infrastructure/OpenIddictProvider.Exchange.cs

```diff
@@ -126,236 +126,283 @@ public partial class OpenIddictProvider<TUser, TApplication, TAuthorization, TSc
             context.Validate();
         }
 
-        public override async Task GrantClientCredentials([NotNull] GrantClientCredentialsContext context) {
+        public override async Task HandleTokenRequest([NotNull] HandleTokenRequestContext context) {
             var services = context.HttpContext.RequestServices.GetRequiredService<OpenIddictServices<TUser, TApplication, TAuthorization, TScope, TToken>>();
 
-            // Retrieve the application details corresponding to the requested client_id.
-            // Note: this call shouldn't return a null instance, but a race condition may occur
-            // if the application was removed after the initial check made by ValidateTokenRequest.
-            var application = await services.Applications.FindByClientIdAsync(context.ClientId);
-            if (application == null) {
-                throw new InvalidOperationException("The token request was aborted because the client application corresponding " +
-                                                   $"to the '{context.ClientId}' identifier was not found in the database.");
-            }
+            Debug.Assert(context.Request.IsAuthorizationCodeGrantType() ||
```

Comment:
Does it make sense to log something if the `grant_type` value is not supported? I.e. remove the Debug.Assert in favor of a proper check?